src/lib/Syntax.hs

```diff
@@ -745,7 +751,10 @@ traverseType :: Applicative m => (Kind -> Type -> m Type) -> Type -> m Type
 traverseType f ty = case ty of
   BaseType _           -> pure ty
   IntRange a b         -> liftA2 IntRange (f depIntType a) (f depIntType b)
-  IndexRange a ac b bc -> liftA4 IndexRange (f depIntType a) (pure ac) (f depIntType b) (pure bc)
+  -- TODO: We really shouldn't be passing depIntType into f in this case
+  IndexRange a b       -> liftA2 IndexRange
+    (traverse (\(t, i) -> liftA2 (,) (f depIntType t) (pure i)) a)
+    (traverse (\(t, i) -> liftA2 (,) (f depIntType t) (pure i)) b)
```

Comment:
This unfortunately ended up looking pretty awful, but it's the first time I've even attempted to write a traversal (I still don't have a good mental model of them), so if you have any ideas as to how to make this code nicer then please do let me know!

Reply:
Yes, these applicative traversals often end up looking ugly. If we really wanted we could try "applicative do" or even just specialize to monads (since we only ever traverse with monads). But this looks fine to me.